test/double-get.c

```diff
@@ -0,0 +1,145 @@
+    }
+
+    /* protect the data */
+    value.data.bo.bytes = NULL;
+    value.data.bo.size  = 0;
+    PMIX_VALUE_DESTRUCT(&value);
+    printf("PMIx_Put on %s\n", key);
+
+
+    return 0;
+}
+
+int pmi_get_string(uint32_t peer_rank, const char *key, void **data_out, size_t *data_size_out)
+{
+    int rc;
+    pmix_proc_t proc;
+    pmix_value_t *pvalue;
+
+    PMIX_PROC_CONSTRUCT(&proc);
+    (void)strncpy(proc.nspace, myproc.nspace, PMIX_MAX_NSLEN);
+    proc.rank = peer_rank;
+    if (PMIX_SUCCESS != (rc = PMIx_Get(&proc, key, NULL, 0, &pvalue))) {
+        ERR("Client ns %s rank %d: PMIx_Get %s: %d\n", myproc.nspace, myproc.rank, key, rc);
+    }
+    if(pvalue->type != PMIX_BYTE_OBJECT){
+        ERR("Client ns %s rank %d: PMIx_Get %s: got wrong data type\n", myproc.nspace, myproc.rank, key);
+    }
+    *data_out = pvalue->data.bo.bytes;
+    *data_size_out = pvalue->data.bo.size;
+
+    /* protect the data */
+    pvalue->data.bo.bytes = NULL;
+    pvalue->data.bo.size = 0;
+    PMIX_VALUE_RELEASE(pvalue);
+    PMIX_PROC_DESTRUCT(&proc);
+
+    printf("PMIx_get %s returned %zi bytes\n", key, data_size_out[0]);
+
+    return 0;
+}
+
+static volatile int fence_active = 0;
+static void fence_status_cb(pmix_status_t status, void *cbdata)
+{
+    fence_active = 0;
+}
+    
+int pmix_exchange(bool flag)
+{
+    int rc;
+    pmix_info_t info;
+
+    fence_active = 1;
+    PMIX_INFO_CONSTRUCT(&info);
+    PMIX_INFO_LOAD(&info, PMIX_COLLECT_DATA, &flag, PMIX_BOOL);
+    if (PMIX_SUCCESS != (rc = PMIx_Fence_nb(&allproc, 1, &info, 1, fence_status_cb, NULL))){
+        fprintf(stderr, "Client ns %s rank %d: PMIx_Fence_nb failed: %d\n", myproc.nspace, myproc.rank, rc);
+        exit(1);
+    }
+    PMIX_INFO_DESTRUCT(&info);
+        
+    /* wait for completion */
+    while(fence_active);
+    
+    return 0;
+}
+
+int main(int argc, char *argv[])
+{
+    char data[256] = {};
+    char *data_out;
+    size_t size_out;
+    int rc;
+    pmix_value_t *pvalue;
+
+    if (PMIX_SUCCESS != (rc = PMIx_Init(&myproc, NULL, 0))) {
+	ERR("PMIx_Init failed");
+        exit(1);
+    }
+    if(myproc.rank == 0) printf("PMIx initialized\n");
+
+    /* job-related info is found in our nspace, assigned to the
+     * wildcard rank as it doesn't relate to a specific rank. Setup
+     * a name to retrieve such values */
+    PMIX_PROC_CONSTRUCT(&allproc);
+    (void)strncpy(allproc.nspace, myproc.nspace, PMIX_MAX_NSLEN);
+    allproc.rank = PMIX_RANK_WILDCARD;
+
+    /* get the number of procs in our job */
+    if (PMIX_SUCCESS != (rc = PMIx_Get(&allproc, PMIX_JOB_SIZE, NULL, 0, &pvalue))) {
+        fprintf(stderr, "Client ns %s rank %d: PMIx_Get job size failed: %d\n", myproc.nspace, myproc.rank, rc);
+        exit(1);
+    }
+    uint32_t nprocs = pvalue->data.uint32;
+    PMIX_VALUE_RELEASE(pvalue);
+
+    /* the below two lines break the subsequent PMIx_Get query on a key set later */
+    pmi_set_string("test-key-1", data, 256);
```

Comment:
@rhc54 If you add `pmix_exchange(true);` in line 131, then the `pmi_get_string` in line 138 fails with error -46